File: skills/ads/scripts/ads_skill/ideas/parse.py
"""Pure markdown-extraction helpers for the /ads:create scaffold: read the
tier-grouped keywords and negative keywords that /ads:gtm authored under
"## Go To Market > ### Keywords". No SDK, no urllib, no stdout, no sys.exit —
every function is referentially transparent and covered by parse_test.py.

The IO shell (bin/create.py) reads the markdown off disk and feeds the text
into these functions; nothing here touches the filesystem or the network."""

from __future__ import annotations

import re

# STAG = Single Theme Ad Group. The scaffold makes one ad group per intent tier
# (Informational/Navigational/Commercial/Transactional) and packs up to this many
# keywords into each. 20 is Google's recommended STAG ceiling. `--top-n` overrides.
DEFAULT_TOP_N = 20
MAX_KEYWORDS_PER_THEME = 20

# STAG themes ARE the intent tiers. The model does the grouping in ads:gtm
# (each keyword is classified into exactly one tier — that classification IS the
# theme assignment). This skill only READS those pre-grouped themes; it makes no
# grouping decision of its own.
_TIER_THEMES = ("Informational", "Navigational", "Commercial", "Transactional")


def _slug_from_processed_path(md_path: str) -> str:
    leaf = md_path.rsplit("/", 1)[-1]
    return re.sub(r"\.(md|markdown)$", "", leaf, flags=re.IGNORECASE)


def _slice_until_next(text: str, pattern: str) -> str:
    m = re.search(pattern, text, flags=re.MULTILINE)
    return text[: m.start()] if m else text


def _strip_offer_suffix(bullet: str) -> str:
    """Drop the ' — offer: ...' / ' -- offer: ...' annotation appended by /ads:gtm
    to multi-intent bullets, leaving just the keyword phrase."""
    return re.sub(r"\s+[—–-]{1,2}\s*offer:.*$", "", bullet, flags=re.IGNORECASE).strip()


def _clean_keyword(raw: str) -> str:
    # Strip /ads:gtm decoration `(volume, competition, $L–$H)` at end.
    stripped = re.sub(r"\s*\([^)]*\)\s*$", "", _strip_offer_suffix(raw))
    return re.sub(r"[*_`]", "", stripped.lower()).strip()


def _extract_keywords(md: str, tiers: tuple[str, ...]) -> list[str]:
    """Pull bullets under the given tier headings (in order), deduped, lowercased,
    offer-suffix stripped. Order: tier order, then bullet order within each tier."""
    gtm = re.search(r"^##\s+Go\s+To\s+Market\b.*$", md, flags=re.IGNORECASE | re.MULTILINE)
    if not gtm:
        return []
    block = _slice_until_next(md[gtm.end():], r"^##\s+")
    kw = re.search(r"^###\s+Keywords\b.*$", block, flags=re.IGNORECASE | re.MULTILINE)
    if not kw:
        return []
    section = block[kw.end():]
    out: list[str] = []
    seen: set[str] = set()
    for heading in tiers:
        m = re.search(rf"^####\s+{re.escape(heading)}\b.*$", section, flags=re.IGNORECASE | re.MULTILINE)
        if not m:
            continue
        sub = _slice_until_next(section[m.end():], r"^####\s+")
        for bullet in re.findall(r"^\s*[-*]\s+(.+?)\s*$", sub, flags=re.MULTILINE):
            cleaned = _clean_keyword(bullet)
            if cleaned and cleaned not in seen:
                seen.add(cleaned)
                out.append(cleaned)
    return out


def _read_theme_groups(md: str, max_per_theme: int) -> list[tuple[str, list[str]]]:
    """Read the Single Theme Ad Groups defined upstream by ads:gtm — one per
    non-empty intent tier, in tier order, keywords in authored order. Truncates a
    theme to `max_per_theme` (Google's per-ad-group ceiling). No grouping logic:
    ads:gtm guarantees each keyword lives in exactly one tier."""
    themes: list[tuple[str, list[str]]] = []
    for tier in _TIER_THEMES:
        kws = _extract_keywords(md, (tier,))[:max_per_theme]
        if kws:
            themes.append((tier, kws))
    return themes


def _extract_negatives(md: str) -> list[dict]:
    """Pull bullets under '#### Negative Keywords' (within ### Keywords) into
    campaign negative-keyword dicts. Phrase only (reason suffix stripped); PHRASE
    match by default. Empty when the section is absent."""
    gtm = re.search(r"^##\s+Go\s+To\s+Market\b.*$", md, flags=re.IGNORECASE | re.MULTILINE)
    if not gtm:
        return []
    block = _slice_until_next(md[gtm.end():], r"^##\s+")
    kw = re.search(r"^###\s+Keywords\b.*$", block, flags=re.IGNORECASE | re.MULTILINE)
    if not kw:
        return []
    section = block[kw.end():]
    m = re.search(r"^####\s+Negative\s+Keywords\b.*$", section, flags=re.IGNORECASE | re.MULTILINE)
    if not m:
        return []
    sub = _slice_until_next(section[m.end():], r"^####\s+")
    out: list[dict] = []
    seen: set[str] = set()
    for bullet in re.findall(r"^\s*[-*]\s+(.+?)\s*$", sub, flags=re.MULTILINE):
        phrase = _clean_keyword(re.split(r"\s+—\s+", bullet)[0])
        if phrase and phrase not in seen:
            seen.add(phrase)
            out.append({"text": phrase, "matchType": "PHRASE"})
    return out

File: skills/adkit/scripts/src/ideas/parse.ts
/**
 * Pure markdown-extraction helpers for the /adkit create scaffold: read the
 * theme-grouped keywords and negative keywords that /adkit gtm authored under
 * "## Go To Market". Ad groups come from the "### Keyword Themes" subsection (one
 * ad group per non-spend-trap theme); negatives come from "### Keywords > ####
 * Negative Keywords". No SDK, no urllib, no stdout, no sys.exit — every function
 * is referentially transparent and covered by parse.test.ts.
 *
 * The IO shell (bin/create) reads the markdown off disk and feeds the text into
 * these functions; nothing here touches the filesystem or the network.
 */

import type { Keyword } from "../lib/schema.js";

// STAG = Single Theme Ad Group. The scaffold makes one ad group per Keyword Theme
// authored by /adkit gtm (step 15c) and packs up to this many keywords into each.
// With 3-6 themes this lands a fresh campaign near the ~100-keyword launch target
// (the gtm generation target, well above the audit's 25-keyword floor); with AI Max
// + Smart Bidding on, more keywords per theme means more data to consolidate, not
// the micro-SKAG anti-pattern. `--top-n` overrides (up to MAX_KEYWORDS_PER_THEME,
// leaving headroom above the default).
export const DEFAULT_TOP_N = 25;
export const MAX_KEYWORDS_PER_THEME = 30;

// A theme flagged by gtm as the generic "keep-but-don't-lead" spend trap. The
// marker may sit anywhere in the `####` heading (before OR after any ` — role
// note`) and is case-insensitive. Such themes are EXCLUDED from ad-group creation
// (they feed the negative-keyword list instead — see reference/gtm.md step 15c).
const SPEND_TRAP_MARKER = /\[\s*spend-?trap\s*\]/i;

/** Escape a string for literal use inside a RegExp (mirrors Python's `re.escape`). */
function escapeRegExp(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

export function slugFromProcessedPath(mdPath: string): string {
  const leaf = mdPath.split("/").pop() ?? mdPath;
  return leaf.replace(/\.(md|markdown)$/i, "");
}

function sliceUntilNext(text: string, pattern: string): string {
  const re = new RegExp(pattern, "m");
  const m = re.exec(text);
  return m ? text.slice(0, m.index) : text;
}

/**
 * The text under a `### <heading>` subsection within the `## Go To Market` block
 * (sliced until the next `###` sibling), or null when either heading is absent.
 * Shared by {@link readThemeGroups} and {@link extractNegatives}.
 */
function gtmSubsection(md: string, heading: string): string | null {
  const gtm = /^##\s+Go\s+To\s+Market\b.*$/im.exec(md);
  if (!gtm) {
    return null;
  }
  const block = sliceUntilNext(md.slice(gtm.index + gtm[0].length), "^##\\s+");
  const re = new RegExp(`^###\\s+${escapeRegExp(heading)}\\b.*$`, "im");
  const m = re.exec(block);
  if (!m) {
    return null;
  }
  return sliceUntilNext(block.slice(m.index + m[0].length), "^###\\s+");
}

/**
 * Drop the ' — offer: ...' / ' -- offer: ...' annotation appended by /adkit gtm
 * to multi-intent bullets, leaving just the keyword phrase.
 */
function stripOfferSuffix(bullet: string): string {
  return bullet.replace(/\s+[—–-]{1,2}\s*offer:.*$/i, "").trim();
}

function cleanKeyword(raw: string): string {
  // Strip /adkit gtm decoration `(volume, competition, $L–$H)` at end.
  const stripped = stripOfferSuffix(raw).replace(/\s*\([^)]*\)\s*$/, "");
  return stripped.replace(/[*_`]/g, "").toLowerCase().trim();
}

/**
 * Pull the group captures of `^\s*[-*]\s+(.+?)\s*$` (MULTILINE) — the bullet
 * bodies — matching Python's `re.findall` with a single capture group.
 */
function bulletBodies(text: string): string[] {
  return [...text.matchAll(/^[ \t]*[-*][ \t]+(.+?)[ \t]*$/gm)].map((m) => m[1]);
}

/**
 * The display name of a `#### ` Keyword Theme heading: the `[spend-trap]` marker
 * and any ` — role note` are removed, leaving the theme's name.
 */
function themeName(heading: string): string {
  return heading
    .replace(SPEND_TRAP_MARKER, "")
    .split(/\s+[—–-]{1,2}\s+/)[0]!
    .replace(/[*_`]/g, "")
    .trim();
}

/**
 * Read the Single Theme Ad Groups authored by ads:gtm under
 * "## Go To Market > ### Keyword Themes" — one `[name, keywords]` pair per
 * `#### ` theme, in file order, `[spend-trap]`-marked themes EXCLUDED (they feed
 * negatives, not ad groups). Keywords are cleaned (decoration/offer suffix
 * stripped, lowercased), deduped ACROSS themes (first-seen wins, so no keyword
 * can land in two ad groups — the no-cannibalization contract), and each theme is
 * truncated to `maxPerTheme` (Google's per-ad-group ceiling). Empty when the
 * `### Keyword Themes` subsection is absent → bin/create dies asking for a gtm
 * re-run (there is deliberately NO fallback to the old intent tiers).
 */
export function readThemeGroups(md: string, maxPerTheme: number): Array<[string, string[]]> {
  const section = gtmSubsection(md, "Keyword Themes");
  if (section === null) {
    return [];
  }
  const headings = [...section.matchAll(/^####\s+(.+?)\s*$/gm)];
  const seen = new Set<string>();
  return headings.flatMap((m, i): Array<[string, string[]]> => {
    const heading = m[1]!;
    if (SPEND_TRAP_MARKER.test(heading)) {
      return []; // spend-trap theme → excluded from ad groups
    }
    const start = (m.index ?? 0) + m[0].length;
    const end = i + 1 < headings.length ? (headings[i + 1]!.index ?? section.length) : section.length;
    const kws = bulletBodies(section.slice(start, end))
      .map(cleanKeyword)
      .filter((c) => c !== "" && !seen.has(c));
    kws.forEach((k) => seen.add(k));
    const capped = kws.slice(0, maxPerTheme);
    return capped.length > 0 ? [[themeName(heading), capped]] : [];
  });
}

/**
 * Pull bullets under '#### Negative Keywords' (within ### Keywords) into
 * campaign negative-keyword objects. Phrase only (reason suffix stripped); PHRASE
 * match by default. Empty when the section is absent.
 */
export function extractNegatives(md: string): Keyword[] {
  const section = gtmSubsection(md, "Keywords");
  if (section === null) {
    return [];
  }
  const m = /^####\s+Negative\s+Keywords\b.*$/im.exec(section);
  if (!m) {
    return [];
  }
  const sub = sliceUntilNext(section.slice(m.index + m[0].length), "^####\\s+");
  const phrases = bulletBodies(sub)
    .map((bullet) => cleanKeyword(bullet.split(/\s+—\s+/)[0]))
    .filter((p) => p !== "");
  return [...new Set(phrases)].map((text): Keyword => ({ text, matchType: "PHRASE" }));
}
